lib/ykpiv.c

```diff
@@ -246,95 +274,119 @@ ykpiv_rc ykpiv_list_readers(ykpiv_state *state, char *readers, size_t *len) {
   return YKPIV_OK;
 }
 
-ykpiv_rc ykpiv_transfer_data(ykpiv_state *state, const unsigned char *templ,
-    const unsigned char *in_data, long in_len,
-    unsigned char *out_data, unsigned long *out_len, int *sw) {
-  const unsigned char *in_ptr = in_data;
-  unsigned long max_out = *out_len;
-  ykpiv_rc res;
+static ykpiv_rc _begin_transaction(ykpiv_state *state) {
   long rc;
-  *out_len = 0;
+  ykpiv_rc res;
 
   rc = SCardBeginTransaction(state->card);
+  if((rc & 0xFFFFFFFF) == SCARD_W_RESET_CARD) {
+    if((res = _reconnect(state)) != YKPIV_OK) {
+      return res;
+    }
+    return _begin_transaction(state);
```

Comment:
isn't a re-call to SCardBeginTransaction() cleaner here to make sure we only try to create the transaction twice?